File: main/src/net/liplum/api/cyber/IStreamClient.java
package net.liplum.api.cyber;

import arc.graphics.Color;
import arc.struct.ObjectSet;
import arc.struct.Seq;
import mindustry.type.Liquid;
import net.liplum.mdt.CalledBySync;
import net.liplum.mdt.ClientOnly;
import net.liplum.common.delegates.Delegate1;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

public interface IStreamClient extends IStreamNode {
    void readStream(@NotNull IStreamHost host, @NotNull Liquid liquid, float amount);

    /**
     * Gets the max acceptable number of this {@code liquid}.
     * negative number means any
     *
     * @param host   host
     * @param liquid liquid
     * @return amount
     */
    float acceptedAmount(@NotNull IStreamHost host, @NotNull Liquid liquid);

    @NotNull
    Delegate1<IStreamClient> getOnRequirementUpdated();
    /**
     * Gets what this client wants<br/>
     * null : Any<br/>
     * Array.Empty : Nothing<br/>
     * An seq : all things in the array<br/>
     * Please cache this value, this is a mutable list.
     * @return what this client wants
     */
    @Nullable
    Seq<Liquid> getRequirements();

    @CalledBySync
    default void connect(@NotNull IStreamHost host) {
        getConnectedHosts().add(host.getBuilding().pos());
    }

    @CalledBySync
    default void disconnect(@NotNull IStreamHost host) {
        getConnectedHosts().remove(host.getBuilding().pos());
    }

    @NotNull
    ObjectSet<Integer> getConnectedHosts();

    default boolean isConnectedWith(@NotNull IStreamHost host) {
        return getConnectedHosts().contains(host.getBuilding().pos());
    }

    /**
     * Gets the maximum limit of connection.<br/>
     * -1 : unlimited
     *
     * @return the maximum of connection
     */
    int maxHostConnection();

    default boolean acceptConnection(@NotNull IStreamHost host) {
        return canHaveMoreHostConnection();
    }

    default boolean canHaveMoreHostConnection() {
        int max = maxHostConnection();
        if (max == -1) {
            return true;
        }
        return getConnectedHosts().size < max;
    }

    default int getHostConnectionNumber() {
        return getConnectedHosts().size;
    }

    @NotNull
    @ClientOnly
    Color getClientColor();
}

File: main/src/net/liplum/api/cyber/IStreamClient.kt
package net.liplum.api.cyber

import arc.graphics.Color
import arc.struct.ObjectSet
import arc.struct.Seq
import mindustry.type.Liquid
import net.liplum.api.ICyberEntity
import net.liplum.common.delegates.Delegate1
import net.liplum.mdt.CalledBySync
import net.liplum.mdt.ClientOnly

interface IStreamClient : ICyberEntity {
    fun readStreamFrom(host: IStreamHost, liquid: Liquid, amount: Float)
    /**
     * Gets the max acceptable number of this `liquid`.
     * negative number means any
     *
     * @param host   host
     * @param liquid liquid
     * @return amount
     */
    fun getAcceptedAmount(host: IStreamHost, liquid: Liquid): Float
    val onRequirementUpdated: Delegate1<IStreamClient>
    /**
     * Gets what this client wants<br></br>
     * null : Any<br></br>
     * Array.Empty : Nothing<br></br>
     * An seq : all things in the array<br></br>
     * Please cache this value, this is a mutable list.
     * @return what this client wants
     */
    val requirements: Seq<Liquid>?
    @CalledBySync
    fun onConnectFrom(host: IStreamHost) {
        connectedHosts.add(host.building.pos())
    }
    @CalledBySync
    fun onDisconnectFrom(host: IStreamHost) {
        connectedHosts.remove(host.building.pos())
    }

    val connectedHosts: ObjectSet<Int>
    fun isConnectedTo(host: IStreamHost): Boolean {
        return connectedHosts.contains(host.building.pos())
    }
    /**
     * Gets the maximum limit of connection.<br></br>
     * -1 : unlimited
     *
     * @return the maximum of connection
     */
    val maxHostConnection: Int
    fun acceptConnectionTo(host: IStreamHost) = canHaveMoreHostConnection
    val canHaveMoreHostConnection: Boolean
        get() = maxHostConnection == -1 || hostConnectionNumber < maxHostConnection
    val hostConnectionNumber: Int
        get() = connectedHosts.size
    @ClientOnly
    val clientColor: Color
}